File: submit/lab5-sol/c/int-vec-add.c
#include <assert.h>
#include <errno.h>
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

typedef struct {
  size_t capacity; //# of ints which can be stored without realloc
  size_t length;   //current # of stored ints
  int *buffer;    //buffer[capacity] with buffer[length] valid
} IntVec;

/** Add value at end of vec */
static void
push_int_vec(IntVec *vec, int value)
{
  enum { INIT_CAPACITY = 2 }; //small init-value for debugging
  if (vec->length >= vec->capacity) {
    assert(vec->length == vec->capacity);
    size_t new_capacity = vec->capacity == 0 ? INIT_CAPACITY : 2*vec->capacity;
    vec->buffer = realloc(vec->buffer, new_capacity*sizeof(int));
    if (vec->buffer == NULL) {
      fprintf(stderr, "cannot realloc vec[%zu]: %s", new_capacity,
              strerror(errno));
      exit(1);
    }
    vec->capacity = new_capacity;
  }
  assert(vec->length < vec->capacity);
  vec->buffer[vec->length++] = value;
}

/** Return IntVec containing all int's read from stdin */
static IntVec read_ints() {
  FILE *in = stdin;
  IntVec ints = {0}; //all fields initialized to 0
  bool is_done = false;
  do {
    int val;
    is_done = (fscanf(in, "%d", &val) != 1);
    if (!is_done) push_int_vec(&ints, val);
  } while (!is_done);
  return ints;
}

/** Write each int in vec on stdout followed by space.  Output a
 *  newline after all the int's have been written.
 */
static void out_int_vec(FILE *out, const IntVec *vec) {
  enum { VAL_WIDTH = 6 };
  for (int i = 0; i < vec->length; i++) {
    fprintf(out, "%d ", vec->buffer[i]);
  }
  fprintf(out, "\n");
}

int main() {
  IntVec ints = read_ints();
  assert(ints.length%2 == 0);

  IntVec vec1 = {0};  //init all fields to zero
  IntVec vec2 = {0};
  IntVec sum = {0};
  for (unsigned i = 0; i < ints.length; i += 2) {
    int val1 = ints.buffer[i], val2 = ints.buffer[i + 1];
    push_int_vec(&vec1, val1);
    push_int_vec(&vec2, val2);
    push_int_vec(&sum, val1 + val2);
  }

  out_int_vec(stdout, &vec1);
  out_int_vec(stdout, &vec2);
  out_int_vec(stdout, &sum);

  //TODO: free buffer's

}

File: submit/lab5-sol/rs/int-vec-add.rs
use std::io::{self, Read};

/** Return Vec<i32> containing all int's read from stdin */
fn read_ints() -> Vec<i32> {
    let mut stdin_str = String::new();
    io::stdin().read_to_string(&mut stdin_str).unwrap();
    let ints: Vec<i32> =  stdin_str
	.split_whitespace()
	.map(|s| s.parse().unwrap())
	.collect();
    ints
}

/** Convert vec into a string having the string representation of
 *  each entry separated by space.
 */
fn vec_str<T: ToString>(vec: &Vec<T>) -> String {
    vec.into_iter()
	.map(|v| v.to_string())
	.collect::<Vec<String>>()
	.join(" ")
}

fn main() {
    let ints = read_ints();
    assert!(ints.len()%2 == 0);
    let mut v1: Vec<i32> = Vec::new();
    let mut v2: Vec<i32> = Vec::new();
    let mut sum: Vec<i32> = Vec::new();
    for i in (0..ints.len()).step_by(2) {
	let i1 = ints[i];
	let i2 = ints[i + 1];
	v1.push(i1); v2.push(i2);
	sum.push(i1 + i2);
    }
    
    println!("{}", vec_str(&v1));
    println!("{}", vec_str(&v2));
    println!("{}", vec_str(&sum));
}
